exp/controllers/azureasomanagedmachinepool_controller.go

```diff
@@ -211,9 +234,56 @@ func (r *AzureASOManagedMachinePoolReconciler) reconcileNormal(ctx context.Conte
 		}
 	}
 
+	agentPool := &asocontainerservicev1.ManagedClustersAgentPool{}
+	err = r.Get(ctx, client.ObjectKey{Namespace: asoManagedMachinePool.Namespace, Name: agentPoolName}, agentPool)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("error getting ManagedClustersAgentPool: %w", err)
+	}
+
+	managedCluster := &asocontainerservicev1.ManagedCluster{}
+	err = r.Get(ctx, client.ObjectKey{Namespace: agentPool.Namespace, Name: agentPool.Owner().Name}, managedCluster)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("error getting ManagedCluster: %w", err)
+	}
+	if managedCluster.Status.NodeResourceGroup == nil {
+		return ctrl.Result{}, nil
+	}
+	rg := *managedCluster.Status.NodeResourceGroup
+
+	clusterClient, err := r.Tracker.GetClient(ctx, util.ObjectKey(cluster))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	nodes := &corev1.NodeList{}
+	err = clusterClient.List(ctx, nodes,
+		client.MatchingLabels(expectedNodeLabels(agentPool.AzureName(), rg)),
```

Comment:
Who is responsible for applying these labels onto nodes? cloud-node-manager (the Azure cloud-provider daemonset)? Or AKS?

Reply:
Basically trying to figure out how CAPZ would handle this behavior changing (if it's a cloud-provider-azure contract then we're good — if it ever changed it'd be tied to a particular release and we could make the change according to that k8s/cloud-provider-azure release; if it's AKS then this is more brittle)

Reply:
I'm not sure exactly who specifies these labels. Even if it is cloud-provider-azure though, is that something we have any control over in an AKS context? We don't install the cloud-provider chart for managed clusters like we do for self-managed.

FWIW this looks like evidence we've had to handle this in the past with our current approach, so I'm not sure there's one obviously proper way to do this: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/f71c5ed8c6ff38a22fe427841f4c527c88cd81d7/controllers/azuremanagedmachinepool_reconciler.go#L137-L145

Reply:
It looks like it's AKS that defines these labels.

Reply:
I would probably vote for another method of durably detecting machine-node affinity. WDYT?

Reply:
Taking another look to see if there's a better way to map a Node to an agent pool, I'm not sure I see anything obviously better though that wouldn't involve querying VMSS, or making assumptions about how AKS names a VMSS: https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-provider-azure-e2e-aks-main/1782867425633505280/artifacts/clusters/capz-e2e-4govhl-aks/nodes/aks-pool0-17336746-vmss000000/node-describe.txt

Reply:
> I would probably vote for another method of durably detecting machine-node affinity. WDYT?

The documentation around these Node labels at least has existed for over two years, so it seems these are relatively stable even if there's no formal guarantee: https://github.com/MicrosoftDocs/azure-docs/commit/0d191135a9c03aca7d23aac1a602a7f91da2244c. I don't think there are any formal guarantees either around the `aks-managed-poolName` tag on a VMSS that CAPZ currently uses to do this mapping, so I'm not sure if one is measurably better than the other.

Overall, I'd vote to stick with the Node labels for now at least while this is experimental. Unless you know of something more reliable?